docs/implementing-new-trigger.md

```diff
@@ -0,0 +1,71 @@
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+description: CRD object for Kafka trigger type
+kind: CustomResourceDefinition
+metadata:
+  name: kafkatriggers.kubeless.io
+spec:
+  group: kubeless.io
+  names:
+    kind: KafkaTrigger
+    plural: kafkatriggers
+    singular: kafkatrigger
+  scope: Namespaced
+  version: v1beta1
+```
+
+Give appropriate and intutive name to the event source.
+
+## Model the CRD spec
+
+Once CRD is defined, you need to model the event source and its attributes as resource object spec. Please see [API conventions] for key attributes of Kubernetes API resource object. Except fot `Spec` part rest of the needed parts to define a Trigger are pretty similar to other Triggers.
+
+For e.g below is the definition of [Kafka Trigger](https://github.com/kubeless/kubeless/blob/master/pkg/apis/kubeless/v1beta1/kafka_trigger.go)
+
+```go
+type KafkaTrigger struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              KafkaTriggerSpec `json:"spec"`
+}
+```
+
+You need to model the event source attributes in to Spec attribute of new trigger. Dependeing on the nature of event source you may want to associate single function or multiple functions with the event source. Use appropriate mechanism to represent the association. For e.g Kafka trigger uses Kubernetes label selector to associate any function with matching label with the event source.
+
+## Code Generation
+
+Once you have definged the new trigger, please ensure its placed in `pkg/apis/kubeless/v1beta1/` path, and update `register.go` go to include new trigger type.
+
+Now you can auto-generate the clientset, lister and informers for the new API resource object as well but running `make update` or `./hack/update-codegen.sh`. Auto-generated clientset, lister and informers comes handy in writing the controller in next step.
+
+## CRD controller
+
+Here is the most important step, i.e. writing controller itself. As far as the skelton of controller goes, it would be pretty similar to existing controllers like Kafka trigger controller, nats trigger controller or http trigger controller. Functionally controller does two important things
+
+- watch Kuberentes API server for CRUD operations on the new trigger object and take appropriate actions.
+- when an event occurs in the event source trigger the associated functions.
+
+Please read the code and logic for existing [controllers](https://github.com/kubeless/kubeless/tree/master/pkg/controller) as a referance.
+
+## Building controller binary and docker image
+
+Ensure you controller is independent binary that can be built from the Makefile. Please follow one of the existing controller [cmd](https://github.com/kubeless/kubeless/tree/master/cmd) as referance. Also ensure there is corresponding `Dockerfile` to build the controller image. Plese see the dockerfile for other trigger controller as a [referance](https://github.com/kubeless/kubeless/tree/master/docker).
```

Comment:
`you controller is independent` -> `your controller is an independent`
`Plese` -> `Please`
`referance` -> `reference`